drivers/ltc4150/ltc4150.c

```diff
@@ -0,0 +1,272 @@
+#else
+    dev->discharged.counter++;
+    dev->discharged.total++;
+#endif
+}
+
+int ltc4150_init(ltc4150_dev_t *dev, const ltc4150_params_t *params)
+{
+    timex_t now;
+
+    if (!dev || !params) {
+        return -EINVAL;
+    }
+
+    memset(dev, 0x00, sizeof(ltc4150_dev_t));
+    dev->interrupt = params->interrupt;
+#ifdef MODULE_LTC4150_BATTERY
+    dev->polarity = params->polarity;
+#endif
+    dev->shutdown = params->shutdown;
+    dev->pulses_per_ah = params->pulses_per_ah;
+
+    if (dev->shutdown != GPIO_UNDEF) {
+        if (gpio_init(dev->shutdown, GPIO_OUT)) {
+            DEBUG("[ltc4150] Failed to initialize shutdown pin");
+            return -EIO;
+        }
+        gpio_set(dev->shutdown);
+    }
+
+#ifdef MODULE_LTC4150_BATTERY
+    if (dev->polarity != GPIO_UNDEF) {
+        if (gpio_init(dev->polarity, GPIO_IN)) {
+            DEBUG("[ltc4150] Failed to initialize polarity pin");
+            return -EIO;
+        }
+    }
+#endif
+
+    xtimer_now_timex(&now);
+    dev->last_update_sec = now.seconds;
+    dev->start_sec = now.seconds;
+
+    if (gpio_init_int(dev->interrupt, GPIO_IN, GPIO_FALLING, pulse_cb, dev)) {
+        DEBUG("[ltc4150] Failed to initialize interrupt pin");
+        return -EIO;
+    }
+
+    DEBUG("[ltc4150] Initialized successfully");
+    return 0;
+}
+
+int ltc4150_deinit(ltc4150_dev_t *dev)
+{
+    if (!dev) {
+        return -EINVAL;
+    }
+
+    gpio_irq_disable(dev->interrupt);
+
+    if (dev->shutdown != GPIO_UNDEF) {
+        gpio_clear(dev->shutdown);
+    }
+
+    return 0;
+}
+
+/**
+ * @brief Convert the raw data (# pulses) acquired by the LTC4150 device to
+ *        charge information in millicoulomb
+ *
+ * @param dev            LTC4150 device the data was received from
+ * @param dest           Store the result here in the following format:
+ *                       dest[0] = used, dest[1] = charged, dest[2] = discharged
+ * @param raw_charged    Number of pulses in charging direction
+ * @param raw_discharged Number of pulses in charging direction
+ */
+static void get_coulomb(const ltc4150_dev_t *dev, int32_t *dest,
+#ifdef MODULE_LTC4150_BATTERY
+                        uint32_t raw_charged,
+#endif
+                        uint32_t raw_discharged)
+{
+    int32_t discharged;
+#ifdef MODULE_LTC4150_BATTERY
+    int32_t charged = raw_charged;
+    charged *= 3600000;
+    charged += dev->pulses_per_ah >> 1;
+    charged /= dev->pulses_per_ah;
+#endif
+
+    discharged = raw_discharged;
+    discharged *= 3600000;
+    discharged += dev->pulses_per_ah >> 1;
+    discharged /= dev->pulses_per_ah;
+
+#ifdef MODULE_LTC4150_BATTERY
+    dest[0] = discharged - charged;
+    dest[1] = charged;
+    dest[2] = discharged;
+#else
+    dest[0] = discharged;
+    dest[1] = 0;
+    dest[2] = discharged;
+#endif
+}
+
+int ltc4150_total(ltc4150_dev_t *dev, int32_t *dest)
+{
+    unsigned irq_flags;
+
+    if (!dest || !dev) {
+        return -EINVAL;
+    }
+
+    irq_flags = irq_disable();
```

Comment:
I actually thought about just doing that. But then it would be required to acquire the mutex from the ISR. I'm not sure about how this is in RIOT, but in general it seemed to me as a bad idea to acquire a mutex in interrupt context.

Reply:
right, I also just saw that mutex lock also disables irqs. hmm.. using messages comes to my mind, but that needs a thread.

Reply:
Using atomic accesses could be used with very careful thought. But maybe that would be too hard to debug and to get right.

Reply:
maybe (parts of) the [pulse counter](https://github.com/RIOT-OS/RIOT/blob/master/drivers/pulse_counter/pulse_counter.c) can be reused

Reply:
Without the functionally to keeping track of the charge transferred within the last minute the pulse counter could be used as it is. But I think keeping that data while trying to reuse code would result in more lines of code.